src/zkevm_specs/evm/execution/calldataload.py

```diff
@@ -0,0 +1,56 @@
+from ..instruction import Instruction, Transition
+from ..opcode import Opcode
+from ..table import RW, CallContextFieldTag, TxContextFieldTag
+from ..util import BufferReaderGadget
+from ...util.param import N_BYTES_WORD
+
+
+def calldataload(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+    instruction.constrain_equal(opcode, Opcode.CALLDATALOAD)
+
+    # offset is the 64-bit offset to start reading 32-bytes from start of calldata.
+    offset = instruction.rlc_to_fq_exact(instruction.stack_pop(), n_bytes=8)
+
+    tx_id = instruction.call_context_lookup(CallContextFieldTag.TxId, RW.Read)
+
+    if instruction.curr.is_root:
+        calldata_length = instruction.tx_context_lookup(tx_id, TxContextFieldTag.CallDataLength)
+        calldata_offset = 0
+    else:
+        calldata_length = instruction.call_context_lookup(CallContextFieldTag.CallDataLength)
+        calldata_offset = instruction.call_context_lookup(CallContextFieldTag.CallDataOffset)
+
+    src_addr = offset + calldata_offset
+    src_addr_end = calldata_length + calldata_offset
+
+    buffer_reader = BufferReaderGadget(
+        instruction, N_BYTES_WORD, src_addr, src_addr_end, N_BYTES_WORD
+    )
+
+    calldata_word = []
+    for idx in range(N_BYTES_WORD):
+        if buffer_reader.read_flag(idx):
+            if instruction.curr.is_root:
+                tx_byte = instruction.tx_calldata_lookup(tx_id, offset + idx)
+                buffer_reader.constrain_byte(idx, tx_byte)
+                calldata_word.append(int(tx_byte))
+            else:
+                mem_byte = instruction.memory_lookup(RW.Read, offset + idx)
```

Comment:
It's not clear to me that this is the right approach.  Shouldn't this be using the `src_addr` instead of `offset` and using caller callID instead of current callID?

Reply:
Yes, I've updated this to be the `src_addr`. Also added another test case where call data offset is non-zero.

> using caller callID instead of current callID
Is this needed? My understanding was that the current call ID is sufficient. But I may be wrong, could you also help reviewing this part @icemelon ?

Reply:
I'll try to write an example to show why I think caller_id must be used:

Imagine the following situation:
from call A we do a call B.  A is the calleer, B is the callee.  When doing the call B, A sets some call data [3,4].  It would look like this:

| Step | Curr. Call | operation                  | curr. mem |
| ---- | ---------- | -------------------------- | --------- |
| 0    | A          | `MSTORE(1, 3)`             | `[]`      |
| 1    | A          | `MSTORE(2, 4)`             | `[0,3]`   |
| 2    | A          | `CALL(_, _, _, 2, 2, ...)` | `[0,3,4]` |
| 3    | B          | `CALLDATALOAD(1)`          | `[]`      |

When entering a new call, we are in a new context so we start with empty memory.  Now in step 3, if we read the current call memory we'll find it empty.  It's in the parent's call (B) memory were we'll find the call data.

Reply:
> It's in the parent's call (B) memory

@ed255 you mean `It's in the parent's call (A) memory` , right?

Reply:
Yes, I made the same error in the `CALLDATACOPY` that uses the callee's call id for the memory read. Could you also fix it in this PR?

Reply:
> > It's in the parent's call (B) memory
> 
> @ed255 you mean `It's in the parent's call (A) memory` , right?

yeah! sorry!  I've edited my original message to fix that.

Reply:
@ed255 I've made the change in the latest commit. Could you review again?